### PR TITLE
product: New GraphQL Commerce_Product schema

### PR DESCRIPTION
**Pull request by tomraithel**
The goal of this PR is to have a unified GraphQL schema for Commerce_Product that makes it easier for consumers to display data for product variants and their related "configurable products".

This is the proposed schema we are trying to achieve:

```graphql
type Product {
    __typename: String!, # __typename
    marketPlaceCode: String! #variants.[sku].baseData.marketPlaceCode
    media: Media, # teaserData.media
    price: ProductPrice, # teaserData.teaserPrice
    loyalty: {
        price: ProductLoyaltyPrice, # teaserData.teaserLoyaltyPriceInfo
        earning: ProductLoyaltyEarning # teaserData.teaserLoyaltyEarningInfo
    },
    title: String!, # variants.[sku].baseData.title
    categories: {
        main: # variants.[sku].baseData.MainCategory with fallback to Categories[0]
        all: # variants.[sku].baseData.Categories
    }
    retailer: {
        id: String, # variants.[sku].retailer.id ?
        title: String, # variants.[sku].retailer.title ?
    }
    brand: {
        id: String, # variants.[sku].brand.id ?
        title: String, # variants.[sku].brand.title ?
    }
    description: String ! #variants.[sku].baseData.description
    meta: {
        description: String !, #variants.[sku].baseData.attributes.getAttribute(key: "metaDescription").label
        title: String !, #variants.[sku].baseData.attributes.getAttribute(key: "metaTitle").label
        keywords: [String !], #variants.[sku].baseData.keywords
    }
    urlSlug: String!, #variants.[sku].baseData.attributes.getAttribute(key: "urlSlug").label
    deliveryOptions: [String!], #variants.[sku].baseData.attributes.getAttribute(key: "deliveryOptions").values
    channel: String!: #getIdentifier.split('-')[2] -> could potentially derived from SP request
    locale: String!: #getIdentifier.split('-')[1] -> could potentially derived from SP request
    limitations: Om3_ProductLimitations_Result! #variants.[sku].productLimitations
    attributes: Attributes #variants.[sku].baseData.attributes

    ... on Commerce_ConfigurableProductWithActiveVariant {
        variantMarketPlaceCode: String!
        activeVariationSelection: [ActiveVariationSelection!]
        variationSelections: [VariationSelection]
        baseConfigurableProduct: Commerce_ConfigurableProduct!
    }

    ... on Commerce_ConfigurableProduct {
        # NO_MATCH for all options
        # use first-best combination for all attributes to determine variantMarketPlaceCode
        variationSelections: [VariationSelection]
    }
};

type VariationSelection {
    code: String!
    label: String!
    options: [VariationSelectionOption]
}

type ActiveVariationSelection {
    attributeLabel: String!
    optionLabel: String!
}

type VariationSelectionOption {
    code: String!
    label: String!
    state: VariationSelectionOptionState!
    variantMarketPlaceCode: String!,
}

enum VariationSelectionOptionState {
    ACTIVE
    MATCH
    NO_MATCH
}
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,11 @@
 *  Added desired time to DeliveryForm
 * GraphQL
     * Updated schema and resolver regarding desired time
-
+    * _Breaking_: New schema for products:
```
Comment (carstendietrich) [inline CHANGELOG.md:7]: please add **product** headline, maybe mention the changes in the product fake service

**Review by carstendietrich — CHANGES_REQUESTED** [review]
Looks good to me so far, would like to see some more comments in GraphQL schema maybe to hint that the title and stuff in Commerce_Product_ActiveVariantProduct are one of the variants. Will have a closer look today.

Didn't we also talk about updating the add to cart mutation? Or should we move that to a separate PR?

File: CHANGELOG.md (hunk 1 above)
Comment (tomraithel) [reply CHANGELOG.md]: done

File: product/interfaces/graphql/product/dto/productdto_activevariantproduct_test.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
+package graphqlproductdto_test
+
+import (
+	"math/big"
+	"testing"
+
+	priceDomain "flamingo.me/flamingo-commerce/v3/price/domain"
+	productDomain "flamingo.me/flamingo-commerce/v3/product/domain"
+	graphqlProductDto "flamingo.me/flamingo-commerce/v3/product/interfaces/graphql/product/dto"
+	"gotest.tools/assert"
```
[the remark this answers is not in the document]
Comment (tomraithel) [reply product/interfaces/graphql/product/dto/productdto_activevariantproduct_test.go]: done

**Comment by tomraithel** [comment]
> Looks good to me so far, would like to see some more comments in GraphQL schema maybe to hint that the title and stuff in Commerce_Product_ActiveVariantProduct are one of the variants. Will have a closer look today.
> 
> Didn't we also talk about updating the add to cart mutation? Or should we move that to a separate PR?

@carstendietrich I´ve updated the comments in the schema file and took care about your other feedback. I think for add-to-cart already accepts two ids, therefore we should have nothing to do there... However if there is something that needs to be changed, we would create another PR

**Comment by tomraithel** [comment]
We noticed, that we didn´t correctly reflect the `VariantVariationAttributesSorting` in the options. I´ve updated the tests and the mapper, to account fo that.

**Review by carstendietrich — APPROVED** [review]
@tomraithel looks good to me, update minor comment style.
Question before I merge, how do I access the attributes of the base configurable when I have a Variant selected?

**Comment by tomraithel** [comment]
> @tomraithel looks good to me, update minor comment style.
> Question before I merge, how do I access the attributes of the base configurable when I have a Variant selected?

@carstendietrich you can still access the original domain product from `productDto.Product`. It is just not exposed to graphql atm because we don´t need it anywhere.
Thanks for your GoDoc adjustments

**Pull request merged**